src/zh/yidan/src/eu/kanade/tachiyomi/extension/zh/yidan/Yidan.kt

```diff
@@ -1,135 +1,348 @@
 import eu.kanade.tachiyomi.source.model.MangasPage
 import eu.kanade.tachiyomi.source.model.Page
 import eu.kanade.tachiyomi.source.model.SChapter
 import eu.kanade.tachiyomi.source.model.SManga
 import eu.kanade.tachiyomi.source.online.HttpSource
+import keiyoushi.utils.firstInstance
 import keiyoushi.utils.getPreferences
+import keiyoushi.utils.parseAs
+import keiyoushi.utils.tryParse
+import kotlinx.serialization.encodeToString
 import kotlinx.serialization.json.Json
-import kotlinx.serialization.json.decodeFromStream
-import okhttp3.Headers
-import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.MediaType.Companion.toMediaType
+import okhttp3.OkHttpClient
 import okhttp3.Request
+import okhttp3.RequestBody
+import okhttp3.RequestBody.Companion.toRequestBody
 import okhttp3.Response
-import rx.Observable
+import okhttp3.ResponseBody.Companion.asResponseBody
+import uy.kohesive.injekt.Injekt
+import uy.kohesive.injekt.api.get
 import uy.kohesive.injekt.injectLazy
+import java.text.SimpleDateFormat
+import java.util.Locale
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
 
 class Yidan : HttpSource(), ConfigurableSource {
     override val name get() = "一耽女孩"
     override val lang get() = "zh"
     override val supportsLatest get() = true
-
-    override val baseUrl: String
-
-    init {
-        val mirrors = MIRRORS
-        val index = getPreferences()
-            .getString(MIRROR_PREF, "0")!!.toInt().coerceAtMost(mirrors.size - 1)
-        baseUrl = "https://" + mirrors[index]
+    private val apiUrl = "https://yd-api.hangtech.cn"
+    override val baseUrl: String = getPreferences().run {
+        val customBaseUrl = getString(PREF_KEY_CUSTOM_HOST, "")
+        if (customBaseUrl.isNullOrEmpty()) {
+            val mirrors = MIRRORS
+            val index = getPreferences()
+                .getString(PREF_KEY_MIRROR, "0")!!.toInt().coerceAtMost(mirrors.size - 1)
+            "https://${mirrors[index]}"
+        } else {
+            customBaseUrl.removeSuffix("/")
+        }
     }
 
-    override fun headersBuilder() = Headers.Builder()
-        .add("User-Agent", System.getProperty("http.agent")!!)
+    override val client: OkHttpClient = super.client.newBuilder().addInterceptor { chain ->
+        val request = chain.request()
+        val response = chain.proceed(request)
+        val requestUrl = request.url.toString()
+        if (requestUrl.contains("images/mhtp/yidan")) {
+            // remove first two bytes for image response
+            val ext = requestUrl.substringAfterLast(".", "png")
+            response.newBuilder().body(
+                response.body.source().apply { skip(2) }
+                    .asResponseBody("image/$ext".toMediaType()),
+            ).build()
+        } else {
+            response
+        }
+    }.build()
 
     private val json: Json by injectLazy()
+    private val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
```

Comment:
Make this static, since the string is not affected by the locale you can use 'no locale':
```suggestion
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.ROOT)
```